### ISSUE TITLE
Wrap contracts with a Javascript library

### ISSUE THREAD

**Issue by hardlydifficult** [unlock-protocol/unlock#1349]
**Is your feature request related to a problem? Please describe.**
This proposal may make it easier to implement breaking changes as we add functionality to our contracts.

**Describe the solution you'd like**
Wrap the smart contracts with a Javascript (or even better, Typescript) library.

The goal is to abstract away as many details as possible so the front-end and contract tests are not so coupled with the specific implementation details.

For example:
 - If we add a parameter to the `createLock` call the process today requires we update several files independently… and this approach does not allow us to define a convenient default.
 - If we change purchasing keys to target Unlock instead of the Lock contract directly, that requires significant updates.  However with a JS library wrapping it, purchasing is exactly the same - it’s still just a Lock address, etc.. the user’s of our JS library API would not need to know that it’s pointing at the Unlock contract now.

This would be a fair bit of work up-front, but may offer long term benefits.  We could use this library to provide default values where appropriate and do type conversion as needed (e.g. BigNumbers!).  This library could also abstract away all details about version management as we go deeper on that topic (well one detail that remains is the library may throw errors if you attempt to use a new feature with an old contract version).

If we go this direction, we would aim to never hit the contract directly - neither the front-end or the contract unit tests.  Everything goes through the library.  By having the contract tests use the library as well, we are thoroughly testing a portion of the javascript logic which may help the front-end team as we make more and more updates.

**Describe alternatives you've considered**

**Additional context**

**Comment by cellog** [comment]
how different would this look from what we have now? Intriguing idea!

**Comment by hardlydifficult** [comment]
I'm not very familiar with how you are doing things today.  Poking around it looks like you are basically doing this already.  It would be similar to the web3Service.js and walletService.js.  The differences I had in mind:

 - Contract unit tests go through the same files (I guess this is the only real difference)
 - Deal with versioning, e.g. maybe a v0 contract throws an error, v1 calls with (x,y) and throws if z was defined, v2 calls with (x,y,z)
 - Maybe drop json dependencies and instead hard code abi per method (less bloat as we add versions)
 - Maybe add caching, e.g. cache Lock address to Lock version to make dealing with versioning faster
 - Deal with BigNumbers consistently, e.g. this line is probably okay but nothing guarantees that it is.  The contract tests recently corrected the same issue. https://github.com/unlock-protocol/unlock/blob/5a3502d0f2d07a1bfbfbfee2f833dd2f670e023e/unlock-app/src/services/web3Service.js#L391 
 - For the contract team: stop dealing with Truffle weirdness!  I personally do not believe this is a good way to test.  Every other user of the system will go through web3, but the tests use Truffle's slightly custom API.  We don't need to.  Let's make the contract tests use web3 too.
 - Fail fast: this library can be responsible for doing things like confirming parameter types and value ranges, and check version compatibility, and then fail instead of relying on an error from the TX failing.
 - Map contract errors to an enum

**Comment by hardlydifficult** [comment]
To clarify Fail fast, the contract tests would still need to confirm contract error codes of course.  The addition of solidity-coverage (test coverage) will help to ensure we don't overlook that concern.

**Comment by cellog** [comment]
I see, so this would be a bigger change on the smart-contracts side, but would create a unified abstraction layer for the contracts in both apps?

**Comment by hardlydifficult** [comment]
Yes, a "unified abstraction layer" is a great summary.  When I first filed this I did not realize how the front-end was already implemented.

**Comment by cellog** [comment]
well I for one think it's a great idea, it would help remove a lot of complexity from the frontend and be much easier to reason through. It is a big change, so perhaps we could implement it in pieces? Not sure how that would work out. Basically we would implement one smart contract function at a time

**Comment by hardlydifficult** [comment]
Yea, that sounds good.  I can take a stab at one method and migrate one Truffle test to use it, then we can debate the approach in PR.

**Comment by cellog** [comment]
I think before you get started, @julien51 should also weigh in

**Comment by nfurfaro** [comment]
I'm interested in this approach and think a Proof of concept PR makes sense as well pending Julien's thoughts.

**Comment by julien51** [comment]
Let me wrap my head around this ;)

**Comment by cellog** [comment]
One thought: we could use a driver pattern, where the main library provides the functions we would call, and we can create drivers that actually implement the calls to the smart contract. This would also allow us to easily migrate from web3 to ethers.js or whatever else comes along, and it would also make it easy to upgrade the contract and even to support multiple contract versions simultaneously.

A decent model of this design pattern is ReactDnD, which is now quite old, but by using drivers, it becomes possible to support both html5 drag and drop and mobile drag and drop with the same interface.

This would have the side benefit of vastly simplifying the web3 and wallet services as well, as we can hide the asynchronous complexity of the actual connections to web3 and company 

Anyways, this strikes me as a big discussion, one to have slowly and deliberately over the next few weeks

**Comment by akeem** [comment]
Definitely jazzed by the idea, as it present a viable ways for people to interact and extend with what we have built this far. 

I do think we should be considerably thoughtful about how and what we expose as it will affect adoption. If we choose correctly, the implementation details shouldn't matter all that much and developers should be able to build whatever they desire on top

**Comment by julien51** [comment]
I agree with the need to "ship" a low level library to wrap calls to the lock. 
This low level library should certainly support versioning and be able to query a lock using its corresponding version.

I am not sure though that we should go as far as actually use that to test our smart contract. I know/understand this is painful with truffle but I want to make sure we have "unit" tests which are as close as possible to the smart contract. (maybe we can consider writing these using solidity or using another framework?).

If I were to summarize things here are the feature I think this library should support:

* Versioning transparently: based on the lock version the library will issue the right query or fail with some kind of "version not supported". 
* data conversion/typing: we define types which are consistently used by the library
* error mapping

Optionally  this library could support caching (either in memory of by connecting to external cache) and invalidation (based on new blocks being added with data changes?).

Now, as much as I understand the benefits, I wonder about the opportunity to do that right now. Our smart contracts are still very much a moving target and I think we should let ourselves evolve fast (even if we break more things on the client side, which is easy to fix). I worry that maintaining this library my end up very time consuming which mean one of 2 things: either we don't maintain this very well or we end up not being able to quickly evolve things. I also think an effort like this would become very valuable in the context of seeing 3rd party dev starting to build on top of the smart contract (using JS too). When that happens we have a lot more context (external at least!) about what the library should do and how it should do this.

TL;DR: great idea... too early.

**Comment by hardlydifficult** [comment]
The driver pattern is what I used with a previous project, and we found it beneficial.  The "Library" was responsible for executing the contract methods and supported multiple blockchains (i.e. web3 and another web3-like option).  To achieve this we simply moved all web3 related logic into this abstraction layer.  This library would help us with migrating to ethers.js or even simply upgrading web3 (which can be breaking at times).

Another example: I have an open PR which changes the params for creating a lock (#1419).  ATM it requires touching a lot, the library would simplify this by allowing something like:

```
function createLock(..., newParam = defaultValue) {...}
```

I understand the concern of moving tests further from the contracts themselves.  I believe the only real implication of this would be around any "fail fast" logic that we add.  Otherwise the library is not much more than a wrapper for web3 calls.

Also yes, we could supplement with Solidity tests and confirm coverage with solidity-coverage.

I also understand the opportunity cost concern.  However if we limit the responsibilities, focusing on providing a wrapper abstracting away web3 dependencies, I think a transition would not be very painful and things would be easier as we make more breaking changes and flush out our versioning strategy.  This is very comparable to what our contract tests are doing today - except our tests include a fair bit of copy-paste and they use truffle APIs instead of web3 APIs.

**Comment by julien51** [comment]
> To achieve this we simply moved all web3 related logic into this abstraction layer. This library would help us with migrating to ethers.js or even simply upgrading web3 (which can be breaking at times).

This is really the logic behind both `web3Service` and `web3Wallet` which are ways to "read" and "write" to smart contracts. 

I would actually be in strong favor of writing a "code" npm module which implements the logic of these.

**Comment by hardlydifficult** [comment]
Yea an npm module for this wrapper code is exactly what we used on a previous project.  I guess the impact of this proposal is 1) create a clear separation for that logic and 2) start migrating the smart-contract tests to use it as well.

RE #2, I did a bit more research on the ZOS update last night.  One of the changes is they just upgraded to Web3 1.0 beta libs, which causes all our test code to break.  If we were using the wrapper logic then we would almost completely bypass their changes. So the ZOS upgrade is an example where we could save time in the long run.

**Comment by cellog** [comment]
Can you describe what you would want from this wrapper? There are a couple of valid approaches. One would be what we have already in the web3Service and walletService, which is a very high-level abstraction. Another would be a very low-level abstraction which allows calling the contract functions with a thin layer over the web3 implementation. And, of course, some kind of hybrid.

What would look good from the smart-contracts end?

**Comment by hardlydifficult** [comment]
At the moment I would like an abstraction layer so we don't use the API provided to us directly.  

Thanks to the ZOS update we now have two unique ways of calling a method - `method.call(param)` or `method(param).call()` and events may appear one of 3 ways `tx.logs[index]` or `tx.events.eventName` or require a follow up call to web3 cause we only have the tx ID.  When providing options, is it `method(param, {options})` or `method(param).send({options})`.  It's getting confusing and I often approach these issues with a simple guess and check approach.

Next best would be to reduce redundancy and some of the issues that result from inconsistencies.  An example of this is BigNumber - we always need to explicitly cast numbers to use bignumber.js since without this the numbers seem to break with every Truffle update.  I've already submitted a number of PRs to fix BigNumber usage.

This PR for the ZOS update introduced a lot of complexity - but the update would have been just a few lines of code if we had an abstraction layer in place already. #1756

We could also consider an abstraction layer for use by the smart contract tests only.

**Comment by cellog** [comment]
ok. So would you say you're looking for a single way to call the method? Or that methods only have 1 valid way to be called, and this is the kind of thing you would need to abstract?

Are you looking for abstracting things like "purchase a key" into a single call? That's what walletService does currently.

**Comment by hardlydifficult** [comment]
Yea - a single way to make a call, casting is done as needed just once, and ideally in a single call (although that's fairly minor I think).  Other nice to haves are deal with versioning (i.e. use the correct ABI) and map errors to an enum instead of using strings directly.

**Comment by cellog** [comment]
ok. I suggest writing a simple abstraction layer for smart-contracts, and then we can look at using it in unlock-app/paywall once we see what you need. Would you use web3 (like `const web3 = new Web3()`) to do the actual calls to the chain?

I would recommend copying some of the constructor from `unlock-app/src/services/web3Service.js` if so

**Comment by hardlydifficult** [comment]
Yes - using Web3 as the wrapper.

I can give it a go, copy-paste web3Service (or part of it) and try migrating one test.  Then we can discuss.

**Comment by julien51** [comment]
@hardlydifficult I started working on something for that purpose in https://github.com/unlock-protocol/unlock/issues/1873
